ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs

```diff
@@ -510,6 +502,14 @@ chainSelectionForBlock cdb@CDB{..} blockCache hdr punish = do
 
     processLoE curChain (LgrDB.ledgerDbCurrent ledgerDB) (writeTVar cdbLoEFrag) cdbLoE
 
+    loeFrag <- atomically (cross curChain <$> readTVar cdbLoEFrag) <&> \case
```

Comment:
`cdbLoEFrag` is now no longer read in the same big STM transaction as everything above. I think this should be fine; in the worst case, we fall back to the immutable tip.

Minor: You could use `readTVarIO` here.